BUG: special: Fix unchecked memory allocations in `specfun.h`

WarrenWeckesser opened the pull request:
A fix for https://github.com/scipy/scipy/issues/21417.

Summary of the changes:

* Add another SF error code, "SF_ERROR_NOMEM", that is used by ufuncs that require internally allocated memory.  Just like the other possible errors, how this error is handled can be queried and set using `geterr()` and `seterr()`.  Unlike the other error conditions, the default behavior is to raise an exception if a memory allocation fails.

* Modify the C++ functions in `specfun.h` that allocate heap memory to return a status code that is used to indicate when the allocation fails.  Modify all callers of such functions to handle the possible error status.

* Fix names passed as the first argument of `set_error()`.  This argument should be the name of the public ufunc, as it will be shown to the user as part of the error message.

Ping @izaid, @ilayn, @steppi.

WarrenWeckesser committed:
WIP: Fix unchecked memory allocation in aswfa()

Summary of the changes:

* Add another SF error code, "SF_ERROR_NOMEM", that is used by ufuncs
  that require internally allocated memory.  Just like the other possible
  errors, how this error is handled can be queried and set using `geterr()`
  and `seterr()`.  Unlike the other error conditions, the default behavior
  is to raise an exception if a memory allocation fails.

* Modify the function `aswfa()` in `specfun.h` to return a status code
  that indicates when a mempory allocation failed.

* Modify the functions `prolate_aswfa_nocv()`, `oblate_aswfa_nocv()`,
  `prolate_aswfa()` and `oblate_aswfa()`  in `sphd_wave.h` to check
  the return value of `aswfa()`, and set the error state and return values
  appropriately if `aswfa()` returns `NoMemory`.

steppi commented on scipy/special/sf_error.pxd:
```diff
@@ -12,6 +12,8 @@ cdef extern from "sf_error.h":
         DOMAIN "SF_ERROR_DOMAIN"
         ARG "SF_ERROR_ARG"
         OTHER "SF_ERROR_OTHER"
+        NOMEM "SF_ERROR_NOMEM"
```
Bikeshed, but I think this would read better as `NO_MEM`. I think `SF_ERROR_MEMORY` fits better with the other names though.

WarrenWeckesser replied:
That sounds good to me. "NOMEM" is a holdover from Posix [`ENOMEM`](https://en.wikipedia.org/wiki/Errno.h#POSIX_errors), but `SF_ERROR_MEMORY` is good too. Any preference on the string that would be used in `geterr()` and `seterr()` (which is public Python API).  I used `"nomem"`, but `"memory"` or `"no_memory"` are also reasonable.

steppi replied:
I like `"memory"`. It seems consistent with `MemoryError` for Python's relevant exception class. I wasn't aware of the Posix symbolic error names but figured  `"nomem"` probably came from somewhere. It was kind of funny reading through the list wondering if the names had a character limit and then finding `ENAMETOOLONG`.

WarrenWeckesser replied:
OK, the new error is "memory".

ilayn commented on scipy/special/xsf/sphd_wave.h:
```diff
@@ -62,14 +62,20 @@ void prolate_aswfa_nocv(T m, T n, T c, T x, T &s1f, T &s1d) {
     }
     specfun::segv(int_m, int_n, c, kd, &cv, eg);
-    specfun::aswfa(x, int_m, int_n, c, kd, cv, &s1f, &s1d);
+    specfun::Status status = specfun::aswfa(x, int_m, int_n, c, kd, cv, &s1f, &s1d);
```
Thanks for this PR @WarrenWeckesser and apologies for this quick-fire question. I might be saying something stupid since it's been a while I looked at these but here is what I was mumbling in the original issue; 

Would it not be better (and possibly higher performant) if we do the malloc calloc here and pass the pointer to aswfa as an extra argument instead of handling it within aswfa? Obviously it will require a change in the function body but then at least it would have the memory error happening here once as it is happening with `segv`? 

I don't know if I am writing it in the right spot of the code but what I am thinking is that the ufunc allocates the memory once and keeps sending it in to aswfa invocation across each element call

steppi replied:
That's what I was thinking too @ilayn. Ideally, I think the memory allocations should happen as high in the call stack they can, under the constraint that we need to know the size of the arrays to be allocated. I'd be willing to restructure the code to make that happen.

WarrenWeckesser replied:
Yes, that is definitely possible.  We could modify the functions in `specfun.h` to require preallocated work arrays.  This wouldn't be an immediate benefit, however, because *this* function (`prolate_aswfa_nocv()`) is still called for every set of scalar inputs, so the same repeated alloc and free is still happening.  To minimize memory allocations, a modification to the ufunc generation infrastructure is needed, so that the memory is allocated before entering the ufunc loop and freed after the loop.  To be specific, the loop that I'm talking about is one of

https://github.com/scipy/scipy/blob/bc0d3271b1243d9c0b3d05f3c2ac67fe4c5eb685/scipy/special/xsf/numpy.h#L819-L838

or

https://github.com/scipy/scipy/blob/bc0d3271b1243d9c0b3d05f3c2ac67fe4c5eb685/scipy/special/xsf/numpy.h#L785-L805

For this example, the `func` that is called in each iteration of the for-loop is `prolate_aswfa_nocv()`.  To minimize the repeated alloc/free, the allocation would need to happen before the for-loop is entered, and the free after the loop is exited.

(This wouldn't eliminate all repeated alloc/free, because when the inputs to the ufunc are higher-dimensional arrays, `loop()` itself is called multiple times by the NumPy ufunc looping machinery.)

This only works if the size of the required work array does not depend on the input values, which is not true for all the functions that do heap allocations.  When the size depends on the value, it might be necessary to do an initial pass over the data that was passed to `loop()` to find the worst case, and allocate that much before entering the main loop.  Or maybe something more clever is possible.

@izaid might have some ideas on how to cleanly incorporate some way of doing this in the C++ code.  Maybe something like pre- and post-loop functions that can be provided in the call that generates a ufunc?  It will need some way to ensure that the arrays that are generated in the pre-loop can be passed to the core function.  Or, perhaps a completely different approach would be that some ufuncs are generated from an explicitly implemented loop function (instead of just the scalar function); then we could do whatever is necessary before and after the loop.

Without such a modification to the ufunc generation code, there is no real benefit to moving the allocation up one level of function call.  Moving the allocation out of the function where it is needed  requires making implementation details of `aswfa()` (for example) available outside the function, which--in general--is something we would like to avoid.  But we definitely want to avoid the repeated alloc/free, so the change is worthwhile, if we anticipate that the required changes to the ufunc generation infrastructure will eventually be made.

steppi replied:
Great thoughts. This sort of thing has been on my mind too, being able to move the allocations up out of the loops, and it's also something Irwin has brought up before. If we can do it, I'm all for it.

For me, the benefit of moving the allocations up in the stack is small but meaningful. It's that it makes it easy to call `set_error` with the actual name of the function, instead of some internal thing like `prolate_awsfa`, without having to pass status codes up through the stack.

WarrenWeckesser replied:
Sorry, I'm not sure I understand your last comment.  Whether the allocation failure is detected in `aswfa()` or in `prolate_aswfa_nocv()` does not affect the name that is passed to `set_error()`.  We can pass whatever we like for the name.  As you can see here, `set_error()` is only called in `prolate_aswfa_nocv()`.  Currently I set the name in `set_error()` to `"prolate_aswfa_nocv"` because that is what the other existing calls to `set_error()` in this function use.  It would probably make sense to change all those to `"pro_ang1"`, since that is the name of the corresponding ufunc in the public Python API, as set here:

https://github.com/scipy/scipy/blob/bc0d3271b1243d9c0b3d05f3c2ac67fe4c5eb685/scipy/special/_special_ufuncs.cpp#L793-L796

ilayn replied:
>  To minimize memory allocations, a modification to the ufunc generation infrastructure is needed, so that the memory is allocated before entering the ufunc loop and freed after the loop. 

Ah OK. Then it looks like a bigger surgery than I initially thought.

steppi replied:
Sorry @WarrenWeckesser. I haven't looked at this code in a while and didn't recall that this function was only called by one other function. I remember in amos, and possiblyy in specfun, there are helper functions which are called by multiple functions, sometimes several layers down in the stack, and I remember thinking it will be annoying to have to propagate the function names to the top if something goes wrong.

In this case, where the function is only called in one place, it's fine to just to change the function name in `set_error1`.

steppi replied:
I briefly talked to @izaid about the idea of hoisting allocations up in the ufunc loops. It's something he's thought about. He wants to be involved in the design process, but is very busy now, so hopes doing this can be put off until he has more time available to think deeply about it. He wants to come up with a general idea which will work for other array library backends as well.

izaid commented:
Just wanted to chime in here. Thanks @WarrenWeckesser for thinking about this. Comments below.

I'm a little concerned we are jumping ahead here. It's not that we don't need better error reporting for memory allocation (we do), it's just that the error handling you see now largely exists because we tried to make something compatible with what was there before. It is totally possible that the error handling you see will totally change in the near future when we separate out `xsf` and properly consider how to do this safely across all array backends.

For that reason, I also don't think we should be modifying the ufunc loops. Whatever you put in the ufunc loops will not translate to other array backends. I also put a lot of effort into getting them right. I'd much prefer we cleanly separate `xsf` into its own library (which @steppi is doing now) and get that in good shape, then worry about error reporting edge cases.

@WarrenWeckesser I really do appreciate the effort you are making, I'd just rather we all sync before you make changes as all of this is very fast moving. Is there some reason you want to do this now in particular? If you want to work on error handling, I'd suggest we discuss how to do it safely across backends.

steppi commented:
@izaid, to clarify, I think this is only incidentally about improving error reporting. @WarrenWeckesser is trying to get rid of unchecked calls to `malloc` so `malloc` failing won't result in a segfault. The reason he's created a new error code is because raising when malloc fails seems like a more sensible default than ignoring it; the default for all other options is to ignore.

ilayn commented:
Also spurious malloc calloc per item is not the way this should work. Memory handling is definitely a sore point in these old functions which is independent from which backend is being used.

WarrenWeckesser committed:
MAINT: Rename 'nomem' -> 'memory'.

WarrenWeckesser commented:
@izaid wrote

>  I'd just rather we all sync before you make changes as all of this is very fast moving. 

Heh, I thought that's what we're doing!  I think there was some confusion over in gh-21417 about the change that I was suggesting, so to make it absolutely clear what I meant, I created this PR, marked "draft", and currently just for one function in `specfun.h`, namely `aswfa()`.  This has been helpful, because I think the discussion has made clear to everyone what we can expect to achieve with a simple change like this.  Specifically, we can fix the unchecked allocations, but to eliminate the alloc/free churn (which is not the goal of this PR), further design work is needed.

Regarding the `scipy.special` error handling (i.e. `geterr()` & `seterr()`): there is no new design here.  I am just adding one new error code that is used to indicate that a memory allocation has failed.  This really should have been added as soon as any ufuncs started using dynamically allocated arrays.  I suspect the original set of error codes (the creation of which predates my time working on SciPy) did not include such an error because the original functions implemented in `special` didn't allocate memory dynamically.  We've had such functions for a while now, so we're overdue for adding the appropriate error status.

So, to be clear, the goal is to fix the immediate problem of unchecked memory allocations.  The changes I am suggesting are pretty straightforward.   If there is agreement, I'm happy to fix the rest of the functions in `specfun.h`.

> Is there some reason you want to do this now in particular?

The problem with unchecked allocations was reported in August.  It would have been nice to fix it for the 1.15 release, but that didn't happen--so it goes, it's software.

steppi commented:
> So, to be clear, the goal is to fix the immediate problem of unchecked memory allocations. The changes I am suggesting are pretty straightforward. If there is agreement, I'm happy to fix the rest of the functions in `specfun.h`.
> 
> > Is there some reason you want to do this now in particular?
> 
> The problem with unchecked allocations was reported in August. It would have been nice to fix it for the 1.15 release, but that didn't happen--so it goes, it's software.

I support moving forward with the general approach in this PR. Thanks @WarrenWeckesser. I guess we can at least try to get this in for the next bug fix release.

izaid commented:
> @izaid wrote
> 
> > I'd just rather we all sync before you make changes as all of this is very fast moving.
> 
> Heh, I thought that's what we're doing! I think there was some confusion over in [gh-21417](https://github.com/scipy/scipy/issues/21417) about the change that I was suggesting, so to make it absolutely clear what I meant, I created this PR, marked "draft", and currently just for one function in `specfun.h`, namely `aswfa()`. This has been helpful, because I think the discussion has made clear to everyone what we can expect to achieve with a simple change like this. Specifically, we can fix the unchecked allocations, but to eliminate the alloc/free churn (which is not the goal of this PR), further design work is needed.
> 
> Regarding the `scipy.special` error handling (i.e. `geterr()` & `seterr()`): there is no new design here. I am just adding one new error code that is used to indicate that a memory allocation has failed. This really should have been added as soon as any ufuncs started using dynamically allocated arrays. I suspect the original set of error codes (the creation of which predates my time working on SciPy) did not include such an error because the original functions implemented in `special` didn't allocate memory dynamically. We've had such functions for a while now, so we're overdue for adding the appropriate error status.
> 
> So, to be clear, the goal is to fix the immediate problem of unchecked memory allocations. The changes I am suggesting are pretty straightforward. If there is agreement, I'm happy to fix the rest of the functions in `specfun.h`.
> 
> > Is there some reason you want to do this now in particular?
> 
> The problem with unchecked allocations was reported in August. It would have been nice to fix it for the 1.15 release, but that didn't happen--so it goes, it's software.

Happy with that, thanks! I think I just saw the comments on the ufunc loop and didn't want us diving into that just yet.

WarrenWeckesser committed:
Fix all uses of unchecked allocations in specfun.h

WarrenWeckesser committed:
Merge branch 'main' into fix-unchecked-allocs

WarrenWeckesser committed:
Fix aswfa() in specfunc.h

WarrenWeckesser committed:
Fix refguide check.

WarrenWeckesser committed:
Add comment about the return value when nonconvergence is detected in rmn2l.

WarrenWeckesser committed:
Change (c/m)alloc calls to use 'new' wrapped in a unique_ptr.

WarrenWeckesser committed:
Fix a few more names passed to set_error()

WarrenWeckesser committed:
Set outputs to nan in mtu12 when alloc fails
Also update docstring-like comments of some C++ functions.

WarrenWeckesser commented:
I updated the PR to fix all the unchecked allocations in `specfun.h`.  I replaced calls of malloc and calloc with the `new (std::nothrow)` operator, and wrapped the pointer in a `std::unique_ptr` so it would be freed automatically when the funcion exits.  This eliminated many explicit calls of `free()`.

steppi commented:
Thanks @WarrenWeckesser.

> I replaced calls of malloc and calloc with the `new (std::nothrow)` operator, and wrapped the pointer in a `std::unique_ptr` so it would be freed automatically when the funcion exits. This eliminated many explicit calls of `free()`.

100% agreed on using `std::unique_ptr` here. RAII saves so much headache for cases like this.

> If this get merged, the commits should be squashed. The opening comment of the PR is probably a sufficient commit message for the single commit.

Sure. I should be able to get to this tomorrow.

steppi reviewed (approved):
Very nice. This looks very thorough. I don't think you've missed anything. I like how you also corrected the function names in the calls to `set_error`. Thanks @WarrenWeckesser!

WarrenWeckesser commented:
Thanks @steppi!

For future reference: in the squashed commit message, the messages from all the commits that were part of the commit history of the PR before squashing could have been removed, and replaced with the opening PR comment.  That's what I had in mind when I suggested "The opening comment of the PR is probably a sufficient commit message for the single commit."  Once the commits are squashed, the commit messages from the old commits aren't really meaningful and just make for a confusing commit message.

ilayn commented:
I think these should go into 1.16 instead of 1.15.2 or am I missing something?

WarrenWeckesser commented:
I'm not against a backport, but I don't think it is urgent; 1.16 should be fine.

The memory management stuff helps ensure that our code handles out-of-memory conditions gracefully, but a user would only encounter an interpreter crash if they happen to call one of the affected functions just as they were running out of memory.

The updated names in the calls to `sf_error` are also probably not significant enough to insist on a backport.

There was one other incidental bug fix that I forgot to mention in a commit message: in the creation of the ufunc `pro_cv`, the wrong name ("obl_cv") was passed to `xsf::numpy::ufunc()`. This was fixed with https://github.com/scipy/scipy/commit/25fb390c163673f688ec7477a310b5d027feeb2d#diff-a4959851a5530860d4aa63d55efcb649dca286df4e450250e366d86da5790428
This is also on the borderline for being backport worthy.

ilayn commented:
Ah then yes, we probably backport it.
